<a href="https://colab.research.google.com/github/leegyuhi/oooooooooooooh/blob/master/Homework_10_8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# __init__.py에 대하여
'__init__.py'의 역할은 파일이 위치한 경로를 패키지 모듈처럼 사용할 수 있도록 해주는 매우 중요한 기능을 수행합니다. 쉽게 말하면 우리가 패키지의 모듈을 추가할때 사용하는 from, import를 이 경로에 사용할 수 있다는 점입니다. 만약 이런 기능이 없다면 모듈을 찾는 경로를 config에 설정해야만 사용할 수 있어 매우 번거롭습니다.

다음과 같은 'Python Files'라는 이름의 패키지로 예를 들어 설명해보겠습니다.

In [0]:
Python Files/
    __init__.py
    myApp.py
    moduleA/
        __init__.py
        getName.py

__init__.py 때문에 getName.py을 import하는 경우 아래처럼 쓸 수 있게됩니다.

In [0]:
from moduleA import getName

위의 moduleA에 위치한 __init__.py에는 공통으로 적용 가능한 기능이나 모듈을 포함할 수 있습니다. 예를들어 해당 경로의 위치한 모든 python 파일들이 동일한 baseAll을 필요로 하는 경우,

In [0]:
! moduleA/a.py
import baseAll

! moduleA/b.py
import baseAll

위와 같이 할 수 있으나 이보단 아래와 같이 두 줄의 코드를 사용할 수 있습니다.

In [0]:
! moduleA/__init__.py
import baseAll

이제 해당 경로의 모든 파일들은 공통으로 baseAll을 사용할 수 있게됩니다.

# 05장 연습문제

## Q1
다음은 Calculator 클래스이다.

In [0]:
class Calculator:
    def __init__(self):
        self.value = 0

    def add(self, val):
        self.value += val

위 클래스를 상속하는 UpgradeCalculator를 만들고 값을 뺄 수 있는 minus 메서드를 추가해 보자. 즉 다음과 같이 동작하는 클래스를 만들어야 한다.

In [6]:
cal = UpgradeCalculator()
cal.add(10)
cal.minus(7)

print(cal.value) # 10에서 7을 뺀 3을 출력

3


In [0]:
# 다음과 같이 Calculator 클래스를 상속하는 UpgradeCalculator 클래스를 만들고 
# minus 메서드를 추가한다.

class UpgradeCalculator(Calculator):
    def minus(self, val):
        self.value -= val

## Q2
객체변수 value가 100 이상의 값은 가질 수 없도록 제한하는 MaxLimitCalculator 클래스를 만들어 보자. 즉 다음과 같이 동작해야 한다.

In [9]:
cal = MaxLimitCalculator()
cal.add(50) # 50 더하기
cal.add(60) # 60 더하기

print(cal.value) # 100 출력

100


단 반드시 다음과 같은 Calculator 클래스를 상속해서 만들어야 한다.



In [0]:
class Calculator:
    def __init__(self):
        self.value = 0

    def add(self, val):
        self.value += val

In [0]:
# Calculator 클래스를 상속하고 add 메서드를 오버라이딩하여 다음과 같은 클래스를 만든다.

class MaxLimitCalculator(Calculator):
    def add(self, val):
        self.value += val
        if self.value > 100:
            self.value = 100

## Q3
다음 결과를 예측해 보자.

하나.

In [0]:
all([1, 2, abs(-3)-3])

둘.



In [0]:
chr(ord('a')) == 'a'

In [10]:
# abs(-3)은 -3의 절댓값이므로 3이 되어 all([1, 2, 0])이 되고, 
# 리스트의 요솟값중 0이 있기 때문에 all 내장 함수의 결과는 False가 된다.

all([1, 2, abs(-3)-3])

False

In [11]:
# ord('a') 의 결과는 97이 되어 chr(97)로 치환된다. 
# chr(97)의 결과는 다시 'a'가 되므로 'a' == 'a'가 되어 True를 돌려준다.

chr(ord('a')) == 'a'

True

## Q4
filter와 lambda를 사용하여 리스트 [1, -2, 3, -5, 8, -3]에서 음수를 모두 제거해 보자.



In [12]:
# 음수를 제거하기 위한 filter의 함수로 lambda 함수를 다음과 같이 만들어 실행한다.

list(filter(lambda x:x>0, [1, -2, 3, -5, 8, -3]))

[1, 3, 8]

## Q5
234라는 10진수의 16진수는 다음과 같이 구할 수 있다.



In [1]:
hex(234)

'0xea'

이번에는 반대로 16진수 문자열 0xea를 10진수로 변경해 보자.

※ 내장 함수 int를 활용해 보자.



In [13]:
int('0xea', 16)

234

## Q6
map과 lambda를 사용하여 [1, 2, 3, 4] 리스트의 각 요솟값에 3이 곱해진 리스트 [3, 6, 9, 12]를 만들어 보자.



In [14]:
# 입력에 항상 3을 곱하여 돌려 주는 lambda 함수를 다음과 같이 만들고
# map과 조합하여 실행한다.

list(map(lambda x:x*3, [1,2,3,4]))

[3, 6, 9, 12]

## Q7
다음 리스트의 최댓값과 최솟값의 합을 구해 보자.



In [0]:
[-8, 2, 7, 5, -3, 5, 0, 1]

In [15]:
# 리스트의 최댓값은 max, 최솟값은 min 내장 함수를 사용하여 다음과 같이 구한다.

a = [-8, 2, 7, 5, -3, 5, 0, 1]
max(a) + min(a)

-1

## Q8
17 / 3의 결과는 다음과 같다.



In [2]:
17 / 3

5.666666666666667

위와 같은 결괏값 5.666666666666667을 소숫점 4자리까지만 반올림하여 표시해 보자.



In [16]:
# round 내장 함수를 사용하면 다음과 같이 반올림하여 소수점 4자리까지 표시할 수 있다.

round(17/3, 4)

5.6667

## Q9
다음과 같이 실행할 때 입력값을 모두 더하여 출력하는 스크립트(C:\doit\myargv.py)를 작성해 보자.

In [0]:
C:\> cd doit
C:\doit> python myargv.py 1 2 3 4 5 6 7 8 9 10
55

※ 외장 함수 sys.argv를 사용해 보자.



In [0]:
!mkdir doit # 'doit' 폴더 생성

In [46]:
# 다음처럼 sys모듈의 argv를 사용하여 명령 행 입력값 모두를 차례로 더해 준다.

%%writefile /content/doit/myargv.py

import sys

numbers = sys.argv[1:] # 파일 이름을 제외한 명령 행의 모든 입력

result = 0
for number in numbers:
    result += int(number)
print(result)

Overwriting /content/doit/myargv.py


In [49]:
%run -i myargv.py 1 2 3 4 5 6 7 8 9 10 # Run the named file inside IPython as a program.

55


## Q10
os 모듈을 사용하여 다음과 같이 동작하도록 코드를 작성해 보자.

1. C:\doit 디렉터리로 이동한다.
2. dir 명령을 실행하고 그 결과를 변수에 담는다.
3. dir 명령의 결과를 출력한다.

In [0]:
# 다음처럼 os 모듈의 chdir을 사용하여 C:\doit 이라는 디렉터리로 이동한다.

import os
os.chdir("/content/doit")

In [0]:
# 그리고 다음처럼 os 모듈의 popen을 사용하여 시스템 명령어인 dir을 수행한다.

result = os.popen("dir")

In [29]:
# popen의 결과를 출력하기 위해 다음과 같이 수행한다.

print(result.read())

myargv.py  __pycache__



## Q11
glob 모듈을 사용하여 C:\doit 디렉터리의 파일 중 확장자가 .py인 파일만 출력하는 프로그램을 작성해 보자.



In [30]:
# 다음과 같이 glob 모듈을 사용한다.

import glob
glob.glob("/content/doit/*.py")

['/content/doit/myargv.py']

## Q12
time 모듈을 사용하여 현재 날짜와 시간을 다음과 같은 형식으로 출력해 보자.



In [0]:
2018/04/03 17:20:32

In [31]:
# time 모듈의 strftime을 사용하여 다음과 같이 작성한다.

import time
time.strftime("%Y/%m/%d %H:%M:%S")   # %Y:년, %m:월, %d:일, %H:시, %M:분, %S:초

'2019/10/09 07:53:50'

## Q13
random 모듈을 사용하여 로또 번호(1~45 사이의 숫자 6개)를 생성해 보자(단 중복된 숫자가 있으면 안 됨).



In [32]:
# random 모듈의 randint를 사용하여 다음과 같이 작성한다.

import random

result = []
while len(result) < 6:
    num = random.randint(1, 45)   # 1부터 45까지의 난수 발생
    if num not in result:
        result.append(num)

print(result)

[44, 1, 17, 39, 35, 42]
